iree/compiler/Dialect/HAL/Conversion/HALToVM/ConvertBufferOps.cpp

```diff
@@ -27,14 +27,20 @@ class BufferLoadOpConversion
       ConversionPatternRewriter &rewriter) const override {
     IREE::HAL::BufferLoadOp::Adaptor adaptor(operands);
     auto importType = importOp.getType();
+    auto resultType = op.getResult().getType();
     auto sizeConst = rewriter.createOrFold<mlir::ConstantOp>(
         op.getLoc(),
         rewriter.getI32IntegerAttr(
             IREE::HAL::getRoundedElementByteWidth(op.getResult().getType())));
-    rewriter.replaceOpWithNewOp<IREE::VM::CallOp>(
-        op, rewriter.getSymbolRefAttr(importOp), importType.getResults(),
+    auto callOp = rewriter.create<IREE::VM::CallOp>(
+        op.getLoc(), rewriter.getSymbolRefAttr(importOp),
+        importType.getResults(),
         ArrayRef<Value>{adaptor.source_buffer(), adaptor.source_offset(),
                         sizeConst});
+    // If the original result was a floating point type, we want to bitcast
+    // from importType (i32) to a matching bit depth floating point type (f32).
+    rewriter.replaceOpWithNewOp<BitcastOp>(
```

Comment:
fyi: if the condition is straightforward it's often worth it to check before emitting the known-foldable case - this reduces the amount of transient IR emitted that requires subsequent transforms to walk it all and clean it up. here, for example, you could do:
```
auto newType = typeConverter->convertType(resultType);
auto callResult = callOp.getResult(0);
if (newType == callResult.getType()) {
  rewriter.replaceOp(op, {callResult});
} else {
  rewriter.replaceOpWithNewOp<BitcastOp>(op, newType, callResult);
}
```

then only if you do load a non-i32 type do you get the bitcast and otherwise no extra IR to deal with.

(this expansion/contraction is what your IR counting tool is helpful to optimize - but best optimization is to not do it in the first place :)

Reply:
Ah, I had something like that before but went for the more elegant code :P

Good point, will update it.